simple-generated/src/event.rs

```diff
@@ -1,87 +1,116 @@
-        let tick_rate = Duration::from_millis(tick_rate);
+    pub fn new(tick_rate: Duration) -> Self {
         let (sender, receiver) = mpsc::channel();
-        let handler = {
-            let sender = sender.clone();
-            thread::spawn(move || {
-                let mut last_tick = Instant::now();
-                loop {
-                    let timeout = tick_rate
-                        .checked_sub(last_tick.elapsed())
-                        .unwrap_or(tick_rate);
-
-                    if event::poll(timeout).expect("failed to poll new events") {
-                        match event::read().expect("unable to read event") {
-                            CrosstermEvent::Key(e) => {
-                                if e.kind == KeyEventKind::Press {
-                                    sender.send(Event::Key(e))
-                                } else {
-                                    Ok(())
-                                }
-                            }
-                            CrosstermEvent::Mouse(e) => sender.send(Event::Mouse(e)),
-                            CrosstermEvent::Resize(w, h) => sender.send(Event::Resize(w, h)),
-                            CrosstermEvent::FocusGained => Ok(()),
-                            CrosstermEvent::FocusLost => Ok(()),
-                            CrosstermEvent::Paste(_) => unimplemented!(),
-                        }
-                        .expect("failed to send terminal event")
-                    }
-
-                    if last_tick.elapsed() >= tick_rate {
-                        sender.send(Event::Tick).expect("failed to send tick event");
-                        last_tick = Instant::now();
-                    }
-                }
-            })
-        };
-        Self {
-            sender,
-            receiver,
-            handler,
-        }
+        thread::spawn(move || event_thread(sender, tick_rate));
+        Self { receiver }
     }
 
     /// Receive the next event from the handler thread.
     ///
-    /// This function will always block the current thread if
-    /// there is no data available and it's possible for more data to be sent.
-    pub fn next(&self) -> AppResult<Event> {
-        Ok(self.receiver.recv()?)
+    /// This function will always block the current thread if there is no data available unless the
+    /// event source has been closed.
+    pub fn next(&self) -> color_eyre::Result<Event> {
+        let event = self.receiver.recv()?;
+        Ok(event)
+    }
+}
+
+/// An event thread that listens for terminal events.
+///
+/// This function is responsible for listening to terminal events and sending them to the main
+/// thread. It sends a tick event at a regular interval. The tick rate is specified by the
+/// `tick_rate` parameter. If an error occurs, it sends an error event to the main thread and then
+/// stops running.
+fn event_thread(sender: mpsc::Sender<Event>, tick_rate: Duration) {
```

Comment:
The order that things happen in this differ a little from the above implementation. This approach should give a tick event immediately, and then on the schedule. The above impl waits for the first timeout to send a tick event.